src/internal/queues.mjs

```diff
@@ -0,0 +1,63 @@
+class QueueItem {
+  constructor (data) {
+    this.data = data
+    this.previous = null
+  }
+}
+
+class Consumer {
+  constructor (queueItem) {
+    this.queueItem = queueItem
+  }
+  isEmpty () {
+    return !this.queueItem.previous
+  }
+  shift () {
+    if (this.isEmpty()) throw new Error('Queue is empty')
+    const data = this.queueItem.previous.data
+    this.queueItem = this.queueItem.previous
+    return data
+  }
+}
+
+export class Queue {
+  constructor () {
+    this.head = new QueueItem() // an empty queue points to a head node
+    this.tail = this.head // initially head and tail are the same
+  }
+
+  push (data) {
+    const newItem = new QueueItem(data)
+    this.tail.previous = newItem
+    this.tail = newItem
+  }
+
+  shift () {
+    if (this.isEmpty()) throw new Error('Queue is empty')
+    const data = this.head.previous.data
+    this.head = this.head.previous
+    return data
+  }
+
+  isEmpty () {
+    return !this.head.previous
+  }
+}
+
+export class Exchange extends Queue {
+  shift () {
+    throw new Error('Unsupported')
+  }
+
+  spawnConsumer () {
+    if (!this.head) throw new Error('You cannot spawn a new consumer after setting calling producerOnly')
+    return new Consumer(this.head)
```

Comment:
OK, so you are also sharing the linked list structure (belonging to the producer) between all the consumers, which grab nodes as needed.

And `this.head` will be the first item unless `noMoreConsumers` is called in which case we can't get here anyway. Alright, makes sense.